Dynamic Knapsack failed the simple test

The simple test is printed out different output:

- [x] FIX THIS AssertionError: 

'<b; [24 chars]\n<c; value: 8.0; weight: 2.0>\n<d; value: 9.0; weight: 5.0>\n' != 
'<b; [24 chars]\n<c; value: 8.0; weight: 2.0>\n'

- [x] pass the simple test

I think this is because the if statement failed to filter the excess baggage d

--- a/lec_2-Optimization/gen_knapsack.py
+++ b/lec_2-Optimization/gen_knapsack.py
@@ -204,3 +204,31 @@ def bruteKnapsack(inputs: list, constraint : float, valueFunction, costFunction)
             opt_list = output_list
     
     return (max_value, opt_list)
+
+def dynamicKnapsack(consider: list, avail: float, taken: tuple = (), val: float = 0) -> list:
+    """  
+    Descrioption: function to optimize knapsack case using dynamic programming
+
+    Parameter:
+    consider : list = list of Thing type objects to be optimized
+    avail : float = the constraint of knapsack available capacity
+    taken : list = list of Thing type objects taken durin optimization (default = [])
+    val : float = value of all taken objects (default = 0)
+    """
+    if taken == None : taken = []
+    if consider == [] :
+        return [consider, avail, taken, val]
+    elif avail < 0 :
+        # set value to be 0 because this should not even considered
+        return [consider, avail, taken, 0]
+    else :
+        consThing = consider[0]
+        nextAvail = avail - consThing.getCost()
+        nextTaken = (*taken, consThing)
+        nextVal = val + consThing.getValue()
+        cons_1, avail_1, taken_1, val_1 = dynamicKnapsack(consider[1:], nextAvail, nextTaken, nextVal)
+        
+        cons_2, avail_2, taken_2, val_2 = dynamicKnapsack(consider[1:], avail, taken, val)
+
+        if val_2 < val_1 : return [cons_1, avail_1, taken_1, val_1]
+        else : return [cons_2, avail_2, taken_2, val_2]

--- a/lec_2-Optimization/gen_knapsack.py
+++ b/lec_2-Optimization/gen_knapsack.py
@@ -212,13 +212,18 @@ def dynamicKnapsack(consider: list, avail: float, taken: tuple = (), val: float 
     Parameter:
     consider : list = list of Thing type objects to be optimized
     avail : float = the constraint of knapsack available capacity
-    taken : list = list of Thing type objects taken durin optimization (default = [])
+    taken : tuple = list of Thing type objects taken durin optimization (default = ())
+    NOTE: I use tuple since it is immutable, if I use list then it will reference the same list 
+          which will return to None or [] each recursive calls!!
     val : float = value of all taken objects (default = 0)
+
+    return : list = [[consideration left over], available_left: float, (optimized Thing), optimized_value: float]
     """
     if taken == None : taken = []
-    if consider == [] :
+    if consider == [] or avail == 0 :
         return [consider, avail, taken, val]
-    elif avail < 0 :
+    elif avail  < consider[0].getCost() :
+        # insufficient available capacity to stored the next thing in the scenario
         # set value to be 0 because this should not even considered
         return [consider, avail, taken, 0]
     else :

--- a/test/test_gen_knapsack.py
+++ b/test/test_gen_knapsack.py
@@ -206,13 +206,16 @@ class TestDynamicKnapsack(unittest.TestCase):
         sys.stdout = cap
 
         # action
-        [cons, avail, taken, val] = gk.dynamicKnapsack(self.things, 5)
+        result = gk.dynamicKnapsack(self.things, 5)
+        # result = [considered: list, avail: float, taken : tuple, val : float]
+        taken = result[-2]
+        val = result[-1]
         for thing in taken:
             print(thing)
         sys.stdout = sys.__stdout__
         printed = cap.getvalue()
 
         # assert
-        self.assertEqual(printed, print_out)
-        self.assertEqual(val, expected_total_value)
+        self.assertEqual(printed, print_out, "printed out is WRONG")
+        self.assertEqual(val, expected_total_value, "TOTAL VALUE is WRONG")
         # self.fail('NO TEST')

--- a/test/test_gen_knapsack.py
+++ b/test/test_gen_knapsack.py
@@ -1,4 +1,4 @@
-import os, sys, unittest
+import os, sys, unittest, io
 
 test_dir = os.path.dirname(__file__)
 code_dir = os.path.normpath(test_dir + "/../lec_2-Optimization")
@@ -183,3 +183,34 @@ class TestBruteKanpsack(unittest.TestCase):
             self.optResult.append(self.subResult)
         
         self.assertEqual(self.optResult, [["clock", 175.0, 10.0], ["painting", 90.0, 9.0], ["book", 10.0, 1.0]])
+
+class TestDynamicKnapsack(unittest.TestCase):
+    """  
+    test for dynamic optimization for knapsack
+    """
+    def setUp(self) -> None:
+        self.datas = {
+            'a' : [6,3],
+            'b' : [7,3],
+            'c' : [8,2],
+            'd' : [9,5]
+        }
+        self.things = gk.buildThings(self.datas, cost_custom = 'weight')
+    
+    def test_simple_things(self):
+        # arrange
+        print_out = '<c; value: 8.0; weight: 3.0>\n<b; value: 7.0; weight: 2.0>\n'
+        expected_total_value = 15.0
+
+        cap = io.StringIO()
+        sys.stdout = cap
+
+        # action
+        [cons, avail, taken, val] = None # TODO: change with the fucntion
+        sys.stdout = sys.__stdout__
+        printed = cap.getvalue()
+
+        # assert
+        self.assertEqual(printed, print_out)
+        self.assertEqual(val, expected_total_value)
+        self.fail('NO TEST')

--- a/test/test_gen_knapsack.py
+++ b/test/test_gen_knapsack.py
@@ -199,18 +199,20 @@ class TestDynamicKnapsack(unittest.TestCase):
     
     def test_simple_things(self):
         # arrange
-        print_out = '<c; value: 8.0; weight: 3.0>\n<b; value: 7.0; weight: 2.0>\n'
+        print_out = '<b; value: 7.0; weight: 3.0>\n<c; value: 8.0; weight: 2.0>\n'
         expected_total_value = 15.0
 
         cap = io.StringIO()
         sys.stdout = cap
 
         # action
-        [cons, avail, taken, val] = None # TODO: change with the fucntion
+        [cons, avail, taken, val] = gk.dynamicKnapsack(self.things, 5)
+        for thing in taken:
+            print(thing)
         sys.stdout = sys.__stdout__
         printed = cap.getvalue()
 
         # assert
         self.assertEqual(printed, print_out)
         self.assertEqual(val, expected_total_value)
-        self.fail('NO TEST')
+        # self.fail('NO TEST')

